### ISSUE TITLE
Security privileges for ESQL Enrich 

### ISSUE THREAD

**Issue by ChrisHegarty** [elastic/elasticsearch#98482]
Filing this issue to track a question/comment that was raised by @ywangd  here https://github.com/elastic/elasticsearch/pull/98309#discussion_r1294115440

One question I do have is why we chose to [execute the LookupRequest with the enrich origin](https://github.com/elastic/elasticsearch/blob/2e4440ff1c0decf79f8a40688566411f7d518aa7/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/EnrichLookupService.java#L141C2-L141C2). It means the following query works when the user has access to only the playlist index without access to either the songs enrich policy or its associated index.

```
FROM playlist | ENRICH songs ON song_id
```

It feels a bit weird to me that we have no control on who can use what enrich process. That said, we don't have a good story around enrich and ingest pipeline either, i.e. we have no control over who can use what pipeline where a pipeline can contain an enrich processor. There have been quite a few complaints from users and we have an https://github.com/elastic/elasticsearch/issues/57586 for it. I wonder whether we should not repeat the problem here with ES|QL. On the surface, it might be easier to tackle compared to ingest pipelines since ES|QL is read-only and has no "bulk" operations. What do you think?

**Comment by ywangd** [comment]
Reposting Nhat's comment from https://github.com/elastic/elasticsearch/pull/98309#discussion_r1295011961 to keep discussion in one place

---
The primary reason we perform the enrich lookup with the enrich origin is to prevent the warning that arises when accessing enrich indices, which are system indices. An alternative is to override allowSystemIndexAccessByDefault in the rest layer, but I felt that it might be too broad. I'm happy to iterate on it.

**Comment by ywangd** [comment]
@dnhatn Accessing system indices with the `enrich` origin is fine. What I am suggesting is to add an additional permission check before switching to the `enrich` origin for system indices. This additional permission check could be something like checking whether the user has permission to access the enrich policy. Here is a sample pseudo code for the idea (please note it's meant to facilitate the discussion, not an implementation porposal).

```java
// This is the outer request that is used to check permission for enrich policy
class EnrichRequest extends ActionRequest {
    private final String enrichPolicyName;
}

class EnrichLookupService {
    void lookupAsync(EnrichRequest enrichRequest, ActionListener listener) {
        // Execute in the end-user context
        transportService.sendChildRequest(.., enrichRequest, ...);
    }

    // This is the transport handler method for EnrichRequest
    void handleEnrich(enrichRequest, listener) {
        // When the code reaches here, we can be sure that the enrich policy is accessible to the end-user.
        // We can now stash to use the enrich origin and execute lookupRequest for system indices access
        ...
    }
}
```

The challenge on the security side is that, other than indices, we don't have a great system to authorize requests for sepcific resources. Cluster actions are authorized based on just the action name and not the specific resource it tries to access. That's part of the reason that we don't have fine-grained authorization for things like pipelines. That said, I think it is still useful even if we could only check whether the end-user has a simple cluster wide access to enrich polices.

**Comment by ywangd** [comment]
@costin This issue got closed by #98641. But I don't see how the two are related. I guess it was an accident? Should I re-open this issue?
